### PR TITLE
skip port reuse for dynamic listeners

### PR DESCRIPTION
**Pull request by sreekanth-ms**
## Description

Prevents dynamically partitioned UDP listeners using the Linux epoll datapath from setting `SO_REUSEPORT` when binding an ephemeral port. Partitioned listeners use a single socket, so port reuse is unnecessary and can allow another process running under the same UID to join the assigned port.

Explicitly bound listeners and the io_uring datapath retain their existing behavior. io_uring support is tracked separately in #6261.

## Testing

Added `DataPathTest.UdpDynamicPortNoReuse` for IPv4 and IPv6 on Linux without io_uring. The test verifies that:

- A dynamic non-partitioned listener still initializes successfully with its existing multi-socket reuse behavior.
- A dynamic partitioned listener receives a non-zero port.
- A second socket using `SO_REUSEPORT` cannot bind that assigned port and fails with `EADDRINUSE`.

## Documentation

No public documentation impact.

### CODE REVIEW

File: src/platform/datapath_iouring.c [hunk 1]
```diff
@@ -897,10 +901,10 @@ CxPlatSocketContextInitialize(
...
 
         //
-        // Only set SO_REUSEPORT on a server socket, otherwise the client could be
-        // assigned a server port (unless it's forcing sharing).
+        // Don't set SO_REUSEPORT for dynamic partitioned listeners.
         //
-        if ((Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL) &&
+        if (!IsDynamicPartitionedListener &&
+            (Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL) &&
             SocketContext->Binding->Datapath->PartitionCount > 1) {
```
[the remark this answers is not in the document]
Comment (sreekanth-ms) [reply src/platform/datapath_iouring.c]: removed the change in iouring for now.

**Comment by leikong** [comment]
@microsoft/quic 
QUIC_SHARED_EPHEMERAL_WORKAROUND seems to be an existing mechanism to fix the "dynamic port resuse" issue within a single process, this PR fixes it for cross-process cases.

**Comment by sreekanth-ms** [comment]
Filed an issue for io_uring https://github.com/microsoft/msquic/issues/6261

**Comment by leikong** [comment]
### io_uring is left on the old behaviour, and the description says otherwise

`src/platform/datapath_iouring.c` still carries the original condition:

```c
//
// Only set SO_REUSEPORT on a server socket, otherwise the client could be
// assigned a server port (unless it's forcing sharing).
//
if ((Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL) &&
    SocketContext->Binding->Datapath->PartitionCount > 1) {
```

The predicate *was* here — added in `1e05c755` ("Align io_uring dynamic port reuse") and removed again by the head commit `9cdf27a0` ("Keep dynamic port fix in epoll").

**Removing it was the right call**, and it is worth recording why, because the obvious follow-up ("just port the predicate") does not work. `CXPLAT_SOCKET_FLAG_PARTITIONED` appears **zero times** in `datapath_iouring.c`. All three behaviours epoll derives from that flag are absent there:

| | epoll | io_uring |
|---|---|---|
| socket count | `!IsPartitioned && PartitionCount > 1` → partitioned gets **1** | `IsServerSocket && PartitionCount > 1` → always **N** |
| partition pinning | `IsPartitioned ? Config->PartitionIndex : (i % N)` | `Config->RemoteAddress ? PartitionIndex : (i % N)` — clients only |
| RSS/BPF steering | `if (!IsPartitioned)` — skipped | `if (IsServerSocket)` — always on |

So a `PARTITIONED` socket on io_uring is an ordinary N-socket server binding that genuinely *needs* `SO_REUSEPORT`; dropping it guarantees `EADDRINUSE` on the second bind. Aligning `SocketCreateUdp` with the epoll version has to come first.

Nothing is broken today: `QUIC_PARAM_LISTENER_PARTITION_INDEX` returns `QUIC_STATUS_NOT_SUPPORTED` under io_uring (`src/core/listener.c`), so the core cannot produce a partitioned listener on that datapath. This is a latent divergence between two Linux datapaths on the same socket-option policy, not a live defect — but it is invisible from the epoll side, which is what makes it worth a note.

**The actionable part:** the PR description still states the check was added to `datapath_iouring.c` and links two diff hunks for it. That half was reverted, so the description now describes code that is not in the diff. Please drop those bullets — otherwise a reader or backporter will believe io_uring is covered.

Worth considering as a follow-up: align io_uring's `SocketCreateUdp`, then apply the same gate there, then drop the `!defined(CXPLAT_USE_IO_URING)` half of the new test's guard so the regression test covers both datapaths.

**Comment by leikong** [comment]
### Prior art: MsQuic already has a workaround for the symptom this PR diagnoses

Worth noting in favour of this change. `src/core/library.c` carries `QUIC_SHARED_EPHEMERAL_WORKAROUND`, compiled in on Linux (`CMakeLists.txt`), whose comment reads:

> *The datapath somehow returned us a "new" ephemeral socket that already matched one of our existing ones. We've seen this on Linux occasionally. This shouldn't happen, but it does.*

That is the same failure this PR addresses, observed from the core layer — and observed often enough to justify a `goto`-based retry loop that walks the port upward until it finds a free one. Note the wording: *"somehow"*, *"this shouldn't happen"*. The symptom was known; the mechanism was not. This PR supplies it — an ephemeral bind carrying `SO_REUSEPORT` is eligible to be placed into an existing reuseport group — and removes the cause for the one socket class where the option was never needed.

Two things follow.

**The workaround only ever caught half of it.** Every check funnels through `QuicLibraryLookupBinding`, which walks `MsQuicLib.Bindings` — this library instance only. A cross-process collision (same UID, both sockets with `SO_REUSEPORT`) produces a **successful** `bind()`, no error anywhere, and nothing to detect. So the retry loop mitigates in-process symptoms; it structurally cannot see the cross-process case. Removing the option at the source, as this PR does, is the only thing that closes it.

**But the retry re-enables the option on the socket this PR just exempted.** When it fires:

```c
SharedEphemeralWorkAround = TRUE;
((CXPLAT_UDP_CONFIG*)UdpConfig)->LocalAddress = &NewLocalAddress;
QuicAddrSetPort((QUIC_ADDR*)UdpConfig->LocalAddress, QuicAddrGetPort(UdpConfig->LocalAddress) + 1);
goto SharedEphemeralRetry;
```

It casts away `const` and rewrites the port to a concrete value, so on the retry the new predicate's `QuicAddrGetPort(...) == 0` clause is false and `SO_REUSEPORT` comes back.

The underlying asymmetry: `QuicLibraryGetBinding` captures `const BOOLEAN PortUnspecified` **before** the `SharedEphemeralRetry:` label precisely so it survives the rewrite — which is why the lookup shortcut needs the extra `&& !SharedEphemeralWorkAround` guard. The core deliberately remembers "the caller asked for any port"; the datapath has no way to, since `CXPLAT_UDP_CONFIG` carries no "stack-assigned rather than caller-chosen" bit.

**Reachability is low and I don't want to overstate it** — entering the retry needs a first-bind collision, which this fix makes essentially impossible for this socket class. The retry becomes dead code *because* the fix works. Not a defect in the PR.

**Suggested disposition:** a line in the description noting it as a known remaining gap. If the predicate ever moves behind an explicit `CXPLAT_UDP_CONFIG` flag set by the caller, this closes for free — the retry rewrites `LocalAddress`, never `Flags`.

**Comment by leikong** [comment]
### `CXPLAT_SOCKET_FLAG_SHARE` says "forces" — for one socket class it's now advisory

`src/inc/quic_datapath.h`:

```c
CXPLAT_SOCKET_FLAG_SHARE = 0x00000002, // Forces sharing of the address and port
```

Every listener sets it, unconditionally — `src/core/listener.c`:

```c
UdpConfig.Flags = CXPLAT_SOCKET_FLAG_SHARE | CXPLAT_SOCKET_SERVER_OWNED; // Listeners always share the binding.
```

`PARTITIONED` is then `|=`'d onto that same config a few lines down. And since `IsDynamicPartitionedListener` requires `RemoteAddress == NULL` — the listener case, per the `CXPLAT_UDP_CONFIG` contract in that same header — **every production socket this predicate can affect has `SHARE` set.** Not an edge case; the entire population.

**To be precise about what does *not* change.** The gate's `SHARE` test is redundant here:

```c
(Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL)
```

Both disjuncts hold for a listener, so `SHARE` was never the proximate cause of `SO_REUSEPORT` on one — `RemoteAddress == NULL` was. Because the predicate itself requires `RemoteAddress == NULL`, it can only subtract from the second disjunct; a client that opted into sharing via `QUIC_PARAM_CONN_SHARE_UDP_BINDING` is untouched. So this is not a logic conflict, and I'm not claiming the gate is wrong.

It's a **contract** conflict: a caller sets a flag whose comment says "forces", and for this socket class it no longer does.

What actually changes for callers:

| | before | after |
|---|---|---|
| in-process join of a dynamic port | already impossible — `QuicLibraryGetBinding` goes to `NewBinding` for any unspecified port | unchanged |
| cross-process join, port learned out-of-band | worked | `EADDRINUSE` |

One scenario: something reads the assigned port back via `CxPlatSocketGetLocalAddress`, publishes it, and another process binds it with `SO_REUSEPORT`. That's the intended casualty — it's the same sharing this PR exists to prevent, and the fix can't distinguish "shared with a stranger" from "shared on purpose after the fact."

**The narrow caveat.** The port-0 rationale — nobody can have arranged to share a port that doesn't exist yet — holds only *before* the first bind. Afterwards the port is knowable, so deliberate sharing becomes expressible, and this forecloses it.

**Ask:** nothing structural, just make the exception visible. Either amend the flag's comment in `quic_datapath.h`, or state it at the predicate. I'd lean toward the header, since that's where the contract is read.

**Comment by leikong** [comment]
### The comment rewrites trade the "why" for the "what"

Two comments changed in this hunk.

**First:**

```diff
-// Only set SO_REUSEPORT on a server socket, otherwise the client could be
-// assigned a server port (unless it's forcing sharing).
+// With multiple datapath partitions, server sockets use SO_REUSEPORT except
+// for dynamic partitioned listeners; clients use it only when sharing.
```

Fair to the rewrite: the old comment never mentioned `PartitionCount > 1`, a third of the condition, and the new one does. Real gain.

But the dropped clause — *"otherwise the client could be assigned a server port"* — was the only statement of **why the condition exists**, and it encodes a non-obvious kernel behaviour:

`SO_REUSEPORT` doesn't just permit an explicit shared bind; it changes what "port in use" means **for the ephemeral allocator**. In Linux's UDP path, a candidate port occupied by a reuseport socket is reported as *available* to another reuseport socket of the same UID — a joinable group, not a conflict. So:

1. A listener holds `0.0.0.0:44380` with `SO_REUSEPORT` — that's the per-processor fan-out working as intended.
2. A client connection starts; `UdpConfig.LocalAddress` is `NULL`, so the datapath binds port 0.
3. If that bind carried `SO_REUSEPORT`, the allocator reaching 44380 sees a reuseport socket of the same UID — same process, even — and calls it available.
4. The client is assigned 44380 and joins the listener's group.

The rule enabling step 3 isn't a defect; it *is* the feature MsQuic relies on for the fan-out. Which is why the guard can't be "make the port unavailable" — it has to be "don't let clients carry the flag," and that is precisely what `RemoteAddress == NULL` does here. `(unless it's forcing sharing)` was the escape hatch for a caller who set `CXPLAT_SOCKET_FLAG_SHARE` deliberately.

What replaced it enumerates which sockets take the branch — which the three lines below already say, in C. The reader now has the rule twice and the hazard zero times. And the hazard is the checkable part: someone widening this condition can test their change against *"could a client end up on a server port?"* They can't test it against a restatement of the code.

**Second:**

```diff
-// The port is shared across processors.
+// Enable port sharing.
```

`Option = TRUE; setsockopt(..., SO_REUSEPORT, ...)` is three lines down; "Enable port sharing" paraphrases it. The old line said something the code doesn't — that the sharing is **across processors**, the fan-out where `SocketCount == CxPlatProcCount()` sockets share one port. That's what the option is *for*, and it's exactly what makes this PR's single-socket exemption make sense.

**Suggestion** — keep the new precision, restore the hazard:

```c
//
// Only set SO_REUSEPORT with multiple datapath partitions, and only on a
// server socket unless the caller forces sharing: an ephemeral bind carrying
// the option can be handed a port an existing reuseport socket of the same
// UID already holds, landing a client on a server's port. Dynamic partitioned
// listeners are excluded -- one socket, so the option buys them nothing while
// exposing them to that same collision.
//
```

For the second, restore "across processors" or drop it. "Enable port sharing" earns nothing.

**While you're in these lines** — two cosmetic notes on the predicate, better done in this pass than as separate comments:

- Declared 287 lines above its only use, wedged between `int Option = 0;` and a zero-init. Moving it beside the gate puts the definition where the question is asked. (Top-of-function `const BOOLEAN` is this file's idiom — but the one in `SocketCreateUdp` is consumed two lines later, not 287.)
- `IsDynamicPartitionedListener` and `IsPartitioned` now coexist over overlapping inputs with near-inverse shapes: `Remote == NULL` is *required* by one and *sufficient to negate* the other. Not contradictory, just a beat of extra work.

I checked one thing before flagging it: the name says "Listener" while the test is `RemoteAddress == NULL` rather than `CXPLAT_SOCKET_SERVER_OWNED`. That's correct as written — the datapath never reads `SERVER_OWNED` anywhere; `RemoteAddress == NULL` is this layer's universal proxy, including in the gate below. No change needed.

---

**Last note, outside the diff and take it or leave it:** `QUIC_SHARED_EPHEMERAL_WORKAROUND` in `src/core/library.c` calls this "the Linux bug where the stack sometimes gives us a 'new' ephemeral port that matches an existing one." That's the same collision from the other side, and it isn't a bug — it's documented `SO_REUSEPORT` behaviour meeting an ephemeral bind. The comment this PR deletes is the one that had it right.

**Comment by leikong** [comment]
### FYI for @microsoft/quic — what this change is for, and the condition behind it

Posting this for the team's awareness rather than as a request. This PR targets `meru-staging` and that's where we're carrying it. What follows is why we needed it and what the surrounding condition looks like — the same condition is on `main`, so the reasoning may be useful to you whether or not anything changes upstream.

`main`'s `datapath_epoll.c` carries the same gate with none of the new predicate:

```c
//
// Only set SO_REUSEPORT on a server socket, otherwise the client could be
// assigned a server port (unless it's forcing sharing).
//
if ((Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL) &&
    SocketContext->Binding->Datapath->PartitionCount > 1) {
```

and everything the new predicate keys on is on `main` too — `CXPLAT_SOCKET_FLAG_PARTITIONED` in `quic_datapath.h`, `listener.c` setting it for a partitioned listener (alongside `SHARE`, unconditionally), and the matching `IsPartitioned` handling in `SocketCreateUdp`. So a dynamic partitioned listener on `main` gets `SO_REUSEPORT` today, and two of them under one UID can converge on a single port.

Nothing below is a change request, against this PR or anywhere else. It's the analysis behind the change, written down because it seems more useful on the record than buried in a staging-branch delta.

#### MsQuic already works around this — but only within a process

This isn't a new concern for the codebase. `QuicLibraryGetBinding` carries an explicit workaround for exactly this phenomenon:

```c
//
// To work around the Linux bug where the stack sometimes gives us a "new"
// empheral port that matches an existing one, we retry, starting from that
// port and incrementing the number until we get one that works.
//
```

It skips the pre-bind lookup for an unspecified port, re-looks-up the *assigned* address after the bind, and on a match traces `BindingError` — "Binding ephemeral port reuse encountered" — tears the binding down and retries at the next port. `QUIC_SHARED_EPHEMERAL_WORKAROUND` is enabled on the Linux branch of the build, which reports *"Enabling shared ephemeral port work around"*. So the phenomenon is already recognised, already considered worth detecting, and already worth carrying code for.

Two things follow from that.

**The detection is structurally process-local.** `QuicLibraryLookupBinding` walks `MsQuicLib.Bindings` — this process's own list. A socket in another process can never appear there, so the identical collision between two processes yields no match, no error and no trace; both sides simply share the port. Same phenomenon, same root cause, and the existing defence cannot be extended to reach it.

**The "Linux bug" is downstream of the option we set.** The allocator hands out a port that is already held only because the holder carries `SO_REUSEPORT`. Measured on Linux — 6000 UDP sockets, each binding `127.0.0.1:0`, all same UID:

| | ports issued more than once |
|---|---|
| `SO_REUSEPORT` | **560** — 610 duplicate assignments, busiest port issued 4× |
| plain | 0 |

Every one of those binds asked for a *new* port. So the workaround is treating a self-inflicted consequence as an external stack defect. Not setting the option where it was never needed removes the cause instead of retrying past the symptom, and for a dynamic partitioned listener leaves the workaround with nothing to catch. It stays load-bearing for an ordinary non-partitioned listener on port 0, whose several sockets genuinely do have to share one port.

#### `SO_REUSEPORT` is a cross-process mechanism gated on a per-process tuning knob

Linux grants the shared bind only when **both** sockets carry the option (same UID). In `datapath_epoll.c` each side's option is gated on its own `Datapath->PartitionCount > 1`:

```c
if (!IsDynamicPartitionedListener &&
    (Config->Flags & CXPLAT_SOCKET_FLAG_SHARE || Config->RemoteAddress == NULL) &&
    SocketContext->Binding->Datapath->PartitionCount > 1) {
```

So whether an unrelated process may share your port is decided by *your* worker-thread count — a precondition that is:

- **unobservable** — neither process can see the other's partition count
- **unnegotiable** — there is no handshake; the failure is a bare `EADDRINUSE`
- **unrelated** — worker count is a throughput knob, not a network policy

A single-vCPU container gets `PartitionCount == 1` and cannot do multi-process port sharing at all. Neither can a process whose `QUIC_EXECUTION_CONFIG` pins it to one processor. Nothing in the API surface hints at either.

It fails symmetrically in both orderings — whichever of the two binds second gets `EADDRINUSE` — and `errno` points at the port, never at the worker count.

#### A global condition, asked about a socket pinned to one partition

`Datapath->PartitionCount` is the worker count for the datapath as a whole. The gate uses it to answer two different questions about one socket, and for a partitioned listener it is a poor fit for both.

*"Do I need this for my own sibling sockets?"* — a within-process resource question. For an ordinary listener it is a fair proxy: `SocketCount` becomes `CxPlatProcCount()` and those sockets really do all bind the same address. But a listener pinned with `SetParam(QUIC_PARAM_LISTENER_PARTITION_INDEX)` is tied to a single partition, and its binding has `SocketCount == 1` — one socket, pinned to `Config->PartitionIndex`. It has no siblings to share with, whatever the datapath-wide worker count happens to be. Nor does another listener in the same process bind alongside it: on an explicit port `QuicLibraryLookupBinding` matches server bindings on local port alone, with no partition awareness, so the second listener shares the existing binding rather than opening a second socket; on port 0 each listener gets its own binding and its own port.

*"Do I permit other processes to share this port?"* — an externally-visible policy question, where a per-process worker count is irrelevant for the reasons above.

So in the partitioned case the honest answer to the first question is "no, by construction", and the second isn't the condition's business at all — yet a global worker count greater than one turns the option on regardless. That is why the behaviour reads as wrong from whichever end you approach it.

#### What the change gives up, checked against the kernel

The change does close something off: a second process joining a port that a dynamic listener was assigned. And the gate's stated rationale — that a client must not be handed a server's port — only bites before the first bind. Both are claims about kernel behaviour, so I measured rather than reasoned (Linux, UDP, same UID; A binds port 0, B binds the port A was assigned):

| A (dynamic bind) | B (explicit bind of A's port) | result |
|---|---|---|
| `SO_REUSEPORT` | `SO_REUSEPORT` | **joins** |
| `SO_REUSEPORT` | plain | `EADDRINUSE` |
| plain | `SO_REUSEPORT` | `EADDRINUSE` |

Port 0 leaves no mark on the resulting socket — it is a request to the allocator and nothing more — so the kernel is content to let someone join afterwards. Row 3 is the post-change state: with A no longer carrying the option, B cannot join however it binds.

What I had not appreciated is that **this costs nothing**. B is a listener too, so B's gate also reduces to `PartitionCount > 1`. Both processes must independently have been given more than one worker, neither can check the other, and the failure is the same bare `EADDRINUSE`. Add the out-of-band channel and the startup ordering — A must publish before B binds, and A picks a different port on restart — and it is an accident that happens to work, not an interface anyone could build on.

So I would not treat this as a trade-off the change needs to defend. The capability was already undependable, and for the same underlying reason.

#### The kernel side of this is the documented contract, not a defect

Worth stating so nobody goes looking for a fix upstream. `socket(7)` defines `SO_REUSEPORT` as permitting "multiple AF_INET or AF_INET6 sockets to be bound to an identical socket address", and the only safeguard it specifies is that all the binding processes share an effective UID — described explicitly as the mechanism that prevents port hijacking. There is no carve-out for an address whose port came from the ephemeral allocator. A port held by a same-UID `SO_REUSEPORT` socket is, by that definition, an address a later socket is entitled to bind, so the allocator offering it follows from the option's stated meaning rather than contradicting it.

Envoy hit this exact problem in its QUIC integration tests — envoyproxy/envoy#19283, *"udp: default-on SO_REUSEPORT doesn't work with listening on port 0"*: multiple instances configured for port 0 under one UID silently converged on a single port, surfacing as test flake. Their reading of the kernel was the same, that it "somewhat aligns with the contract of binding port 0 and `SO_REUSEPORT`", and the production remedy they recommend is the one this PR implements — explicitly disable the option for port-0 listeners. That issue is still open and marked `help wanted`, because the general case, wanting port 0 *and* deliberate sharing, has no clean answer while the application cannot say which it means.

One limit on the above: I did not find a kernel-side report or patch declaring the behaviour either intended or defective. This is the documented contract plus one peer project's reading of it, not an upstream ruling.

#### Why `SHARE` is not the lever it looks like

`SHARE` looks like the natural place for a caller to say whether sharing is wanted, but `src/core/listener.c` sets it on every listener unconditionally. A constant carries no information, so for listeners that clause of the gate is inert, and the datapath is left inferring policy from mechanism — `RemoteAddress == NULL` for "is a server", `PartitionCount > 1` for "might have siblings". The new term is a third inference of the same kind, and it is the best available given the inputs.

#### Scope: this is Linux-only

Worth stating so the team doesn't have to check. `datapath_win*.c` sets no port-sharing socket option at all — no `SO_REUSEADDR`, `SO_PORT_SCALABILITY` or `SO_REUSE_UNICASTPORT` — and its one structurally similar `Config->RemoteAddress == NULL && Datapath->PartitionCount > 1` test drives `SIO_CPU_AFFINITY`, which is within-process. So there is no equivalent cross-process coupling on Windows and nothing to mirror there.

#### Where this leaves things

Summarising the part that's probably most useful to the team:

- Anything that opens a **partitioned listener on a dynamic port** is exposed, and the ephemeral-reuse workaround already in the tree cannot see it, because the colliding socket is in another process. That's the exposure this change removes, and it's why we needed it.
- **It does give something up, and we think that's fine.** A second process can no longer join a port that a dynamic listener was assigned and then published out of band — the kernel does permit exactly that today, per the first row of the table above. But the arrangement needs an out-of-band channel, a start ordering, a fresh port whenever the first process restarts, and both sides independently having been given more than one worker. That is awkward enough to be an accident that happens to work rather than something to build on, so it seems a fair trade for closing the collision.
- We're taking the change on `meru-staging` as written. Separately we're tracking the sturdier form: set a flag in `listener.c` from `PortUnspecified` and test that in the datapath, instead of re-deriving the intent from the address. That version also survives the `QUIC_SHARED_EPHEMERAL_WORKAROUND` retry described above, which the predicate does not.

One question if anyone happens to know the history: **is the `PartitionCount`-gates-cross-process-sharing coupling known and accepted?** If it's a deliberate trade, the first half of this note is just documentation. If it's news, it may be worth a look on its own merits.

**Comment by leikong** [comment]
### The predicate has three clauses; the test supplies one input vector

`IsDynamicPartitionedListener` in `datapath_epoll.c`:

```c
Config->RemoteAddress == NULL &&                                              // (1)
(Config->Flags & CXPLAT_SOCKET_FLAG_PARTITIONED) != 0 &&                      // (2)
(Config->LocalAddress == NULL || QuicAddrGetPort(Config->LocalAddress) == 0); // (3)
```

`UdpDynamicPortNoReuse` supplies exactly one vector: no remote address, `PARTITIONED` set, and `GetNewLocalAddr(false)` yielding a non-null address with port 0 — so (3) is satisfied through its second disjunct only. All three true, one row of the table.

**Clause (1) is doing the most work here, and has no coverage.** `PARTITIONED` has two setters. `listener.c` sets it alongside `RemoteAddress = NULL`. `connection.c` sets it on connection sockets, which always carry `&Path->Route.RemoteAddress` — and which set `LocalAddress` to NULL whenever the app didn't pin one. So for a shared, partitioned connection started without an explicit local address:

- (2) `PARTITIONED` — true
- (3) `LocalAddress == NULL` — true
- (1) `RemoteAddress == NULL` — **false, and that is the only reason the predicate doesn't fire**

`connection.c` also sets `SHARE` when `ShareBinding` is on, so those sockets satisfy the old gate's `SHARE ||` arm and do receive `SO_REUSEPORT` today. Drop clause (1) and they silently stop — a behaviour change to `QUIC_PARAM_CONN_SHARE_UDP_BINDING` resting on a single untested term.

**Clause (3) is never falsified.** A partitioned listener on an explicit non-zero port must keep the option. Nothing exercises that, and nothing else would notice either: a partitioned binding has `SocketCount == 1`, so removing the port check breaks no bind anywhere.

**Clause (2) is the only one with protection, and it's incidental.** Delete it and every ordinary port-0 listener loses `SO_REUSEPORT`, so its per-processor sockets can no longer share the assigned port and *unrelated* tests fail at bind. The test does construct the false case — `NonPartitionedSocket` — but never asserts anything about it, so that protection comes from elsewhere in the suite rather than from here.

**The `LocalAddress == NULL` disjunct is a different category, not a coverage gap.** No `src/core/` caller can produce it together with `PARTITIONED`: listeners always pass `&BindingLocalAddress`, and the one caller that does pass NULL is excluded by clause (1). It's reachable only by hand-building a `CXPLAT_UDP_CONFIG`. Defensible as breadth in the datapath's own contract — noting it only so nobody reads it as covered behaviour.

The shape of it: the test asserts a probe cannot join the port, which is equally true of a build that suppressed `SO_REUSEPORT` unconditionally. It shows the suppression **happens**; it doesn't show it is **narrow**, which is the entire content of writing three clauses instead of one.

**Ask:** two cases, both cheap.

1. Same body, socket built *without* `PARTITIONED`, assert the probe **succeeds**. Converts "suppression occurs" into "suppression is conditional", and gives `NonPartitionedSocket` a purpose.
2. Socket built with a remote address plus `PARTITIONED | SHARE` on port 0, assert the probe **succeeds** — pins clause (1). The `CxPlatSocket` helper already takes a remote-address parameter, so no new scaffolding. (I haven't run this one; a connected UDP socket may need care in the probe.)

**Comment by leikong** [comment]
### Correction to my previous comment

Two claims I made about `NonPartitionedSocket` are wrong. I wrote:

> The test does construct the false case — `NonPartitionedSocket` — but never asserts anything about it, so that protection comes from elsewhere in the suite rather than from here.

and, in ask #1, that adding a negative case would give `NonPartitionedSocket` "a purpose". It already has one, and the protection is right here in this test.

**Why.** Because that socket is non-partitioned on port 0, `SocketCreateUdp` computes `NumPerProcessorSockets` true and `SocketCount = CxPlatProcCount()`. Socket[0] binds port 0 and `getsockname` writes the assigned port back into `Binding->LocalAddress`; sockets[1..N-1] then bind *that* port. That only succeeds with `SO_REUSEPORT` set — so `VERIFY_QUIC_SUCCESS` on its init status is an implicit but load-bearing assertion that the predicate does not fire for an ordinary port-0 listener.

**Verified rather than argued.** Deleting the `PARTITIONED` term from the predicate and rebuilding:

```
DataPathTest.cpp:769: Failure
Value of: (static_cast<int>(NonPartitionedSocket.GetInitStatus()) <= 0)
  Actual: false
Expected: true
[  FAILED  ] DataPathTest/DataPathTest.UdpDynamicPortNoReuse/4
[  FAILED  ] DataPathTest/DataPathTest.UdpDynamicPortNoReuse/6
```

Both parameterizations fail at line 769 — the `NonPartitionedSocket` init check — not at the probe assertions. Pristine build passes 2/2, mutated build fails 2/2 at that line, reverted build passes 2/2 again with the source byte-identical.

**What survives of the observation**, and it's weaker than what I posted: the assertion is invisible locally. What makes it load-bearing runs through `SocketCount`, the two-phase bind and the port-0 write-back, none of which is apparent at the call site — which is why I misread it as inert.

Ask #1 still stands, but as *make the existing check explicit and outward-facing* rather than *give the socket a purpose*: construction succeeding proves the option was set for siblings inside the process; a probe succeeding proves it from outside, in one line, where a reader can see it.

The clause (1) and clause (3) analysis and both asks are unaffected.

**Review by leikong — APPROVED** [review]

**Comment by guhetier** [comment]
Hi, as far as I understand through a quick read, this ties the use of port reuse to an unrelated concept (partitioned listener).
Unless there is a reason that makes port reuse useless with partitioned listeners, I don't think this is a change we would accept in main (for this PR to meru-staging, it is up to you).

But I would suggest trying to understand if there is a solution for this in the Linux ecosystem (like port reservations in Windows).

**Comment by leikong** [comment]
> Hi, as far as I understand through a quick read, this ties the use of port reuse to an unrelated concept (partitioned listener). Unless there is a reason that makes port reuse useless with partitioned listeners, I don't think this is a change we would accept in main (for this PR to meru-staging, it is up to you).
> 
> But I would suggest trying to understand if there is a solution for this in the Linux ecosystem (like port reservations in Windows).

Just in case you hadn't read this comment, details are here:
https://github.com/microsoft/msquic/pull/6251#issuecomment-5414282389
